### PR TITLE
LUCENE-10026: Fix CombinedFieldQuery equals and hashCode

### PR DESCRIPTION
**Pull request by jtibshirani**
The previous equals and hashCode methods only compared query terms. This meant
that queries on different fields, or with different field weights, were
considered the same.

During boolean query rewrites, duplicate clauses are removed. So because
equals/ hashCode was incorrect, rewrites could accidentally drop
CombinedFieldQuery clauses.

### CODE REVIEW

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/CombinedFieldQuery.java [hunk 1]
```diff
@@ -212,13 +226,18 @@ public String toString(String field) {
   }
 
   @Override
-  public int hashCode() {
-    return 31 * classHash() + Arrays.hashCode(terms);
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
```
Comment (jpountz) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/CombinedFieldQuery.java:231]: use `Query#sameClassAs` ?

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/CombinedFieldQuery.java [hunk 2]
```diff
@@ -212,13 +226,18 @@ public String toString(String field) {
...
+    CombinedFieldQuery that = (CombinedFieldQuery) o;
+    return Objects.equals(fieldAndWeights, that.fieldAndWeights) && Arrays.equals(terms, that.terms);
   }
 
   @Override
-  public boolean equals(Object other) {
-    return sameClassAs(other) && Arrays.equals(terms, ((CombinedFieldQuery) other).terms);
+  public int hashCode() {
+    int result = Objects.hash(fieldAndWeights);
+    result = 31 * result + Arrays.hashCode(terms);
```
Comment (jpountz) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/CombinedFieldQuery.java:239]: use `classHash()` in the hashcode?

**Review by jpountz — APPROVED** [review]
LGTM

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/CombinedFieldQuery.java (hunk 1 above)
Comment (jtibshirani) [reply lucene/sandbox/src/java/org/apache/lucene/sandbox/search/CombinedFieldQuery.java:231]: Thanks, I had just auto-generated these.

**Review by jpountz — APPROVED** [review]

**Pull request merged**